Fix: #1655 - Added working Device Check API that is compliant with the DeviceCheck Server

Brandon-T opened the pull request:
- Security Review: https://github.com/brave/security/issues/13
- Spec: https://github.com/brave-intl/bat-captcha/pull/45
- Added working Device Check API that is compliant with the DeviceCheck Server
- Added PEM formatting (ASN.1 DER - ANSI X9.62) for Elliptic Curve Keys following the RFC spec.
- Removed biometric requirements from the Keys.
- Added signing and verifying of challenge nonce with the key as per Brave spec.
- Added example flow for Device Check from Brave spec.
- Made URLs compliant with Brave server instead of RFC spec.

**Hello World:**
```swift
//Sample Flow:
//generate Token -> generateEnrollment -> registerDevice
//getAttestation -> setAttestation (verify it)
// Verify if we failed or not.. and do something with the result..

class DeviceCheckExample {
  private let client = DeviceCheckClient()
  private var token = ""
  private let paymentId = UUID().uuidString //This comes from Brave Rewards..
  private var completion: ((_ error: Error?) -> Void)?
  
  // Starts the device check flow with the server
  func startHandshake(_ completion: @escaping (Error?) -> Void) {
    self.completion = completion
    self.generateToken()
  }

  deinit { self.completion = nil }
  
  // Generates a device check token using the iOS Apple APIs.
  // I do not store this token anywhere atm.
  private func generateToken() {
    client.generateToken { [weak self] token, error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.token = token
      self.generateEnrollment()
    }
  }
  
  // Generates an enrollment structure using the token and making a call to the server.
  private func generateEnrollment() {
    if token.isEmpty {
      self.completion?("Invalid Device Check Token")
      return
    }
    
    client.generateEnrollment(paymentId: paymentId, token: token) { [weak self] registration, error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.registerDevice(registration)
    }
  }
  
  // Registers the device with the server
  private func registerDevice(_ registration: DeviceCheckRegistration?) {
    guard let registration = registration else {
      self.completion?("Error registering device - Unknown Reason")
      return
    }
    
    client.registerDevice(enrollment: registration) { [weak self] error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.getAttestation()
    }
  }
  
  // Retrieves a challenge-nonce from the server
  private func getAttestation() {
    client.getAttestation(paymentId: paymentId) { [weak self] attestation, error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.setAttestation(attestation)
    }
  }
  
  // Responds to the challenge and sends it to the server.
  private func setAttestation(_ attestation: AttestationBlob?) {
    guard let attestation = attestation else {
      self.completion?("Error retrieving attestation - Unknown Reason")
      return
    }
    
    client.setAttestation(nonce: attestation.nonce) { [weak self] error in
      guard let self = self else { return }
      
      if let error = error {
        self.completion?(error)
        return
      }
      
      self.completion?(nil)
    }
  }
}
```

**Hello World Usage:**
```swift
DeviceCheckExample().startHandshake { error in
  if let error = error {
    log.error(error) //Failed to retrieve attestation for this device.. Do something..
    return
  }
  
  //Attestation has been successful so do something else now.. claim the grant I guess?..
}
```

<!-- *Thank you for submitting a pull request, your contributions are greatly appreciated!* -->

## Summary of Changes

This pull request fixes issue https://github.com/brave/brave-ios/issues/1655
<!-- If no ticket has been fixed, please file one now: https://github.com/brave/brave-ios/issues -->

## Submitter Checklist:

- [ ] *Unit Tests* are updated to cover new or changed functionality
- [x] User-facing strings use `NSLocalizableString()`

## Test Plan:
<!-- Any useful notes explaining how best to test and verify. -->


## Screenshots:
<!-- If your patch includes user interface changes that you would like to suggest or that you would like UX to look at, please include them here. -->


## Reviewer Checklist:

- [ ] Issues include necessary QA labels:
  - `QA/(Yes|No)`
  - `release-notes/(include|exclude)`
  - `bug` / `enhancement`
- [ ] Necessary [security reviews](https://github.com/brave/security/issues/new/choose) have taken place.
- [ ] Adequate unit test coverage exists to prevent regressions.
- [ ] Adequate test plan exists for QA to validate (if applicable).
- [ ] Issue is assigned to a milestone (should happen at merge time).

Brandon-T committed:
Added working Device Check API that is compliant with the DeviceCheck server.
Added PEM formatting for Elliptic Curve Keys following the RFC spec.
Removed biometric requirements from the Keys.
Added signing and verifying of challenge nonce with the key as per Brave spec.
Added example flow for Device Check from Brave spec.

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+    var result: CFTypeRef?
+    let error = SecItemCopyMatching(query as CFDictionary, &result)
+    if error == errSecSuccess || error == errSecInteractionNotAllowed {
+      if let result = result as? [String: Any], let accessControl = result[kSecAttrAccessControl as String] as! SecAccessControl? { //swiftlint:disable:this force_cast
```
Do you think it's better to crash the app here instead of unwrapping this and return false?

Brandon-T replied:
![image](https://user-images.githubusercontent.com/1530031/67502936-bf60f900-f654-11e9-8d30-47718b0754d2.png)

<img width="742" alt="Screen Shot 2019-10-24 at 11 51 58 AM" src="https://user-images.githubusercontent.com/1530031/67502939-c25be980-f654-11e9-9d38-fa9766ca1287.png">

This is a bug in Xcode and it can't be done any other way. There's no way to unwrap a `CFTypeRef?` as another `CF` type. You will always end up with a force somewhere :(

I've tried to `if let`, I have tried `guard let`, I have tried optional cast and ternary unwrap.. None of them work and will give the above errors. That's why I have to force cast it to an optional type and then unwrap that optional type :(

iccub replied:
Lol XCode really doesn't like you :D
Would be good to put a comment about it somewhere in code

Brandon-T replied:
I fixed it by first checking what type was returned via its TypeId and then force-casting. This should provide the safety we're looking for.

jumde commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+  }
+  
+  /// Returns the private key
+  public func getPrivateKey() -> SecKey {
```
Do you expect the private key to be used outside this file? If not we can make it private and only expose the utility functions like `sign` and `getPublicKey`.

Brandon-T replied:
True. I will fix it. The private key is not needed outside this file but also, the private key can't be exported as well so it doesn't hurt :)

But yeah, good point. It will be fixed.

Brandon-T committed:
Changed access modifier of public and private `SecKey` as per feedback.

jumde commented on Client/DeviceCheck/DeviceCheck.swift:
```diff
@@ -0,0 +1,367 @@
+
+private extension DeviceCheckClient {
+  // The base URL of the server
+  private static let baseURL = URL(string: "http://reputation-alb-1044480833.us-west-2.elb.amazonaws.com")
```
https

Brandon-T replied:
I got the url from slack:

![image](https://user-images.githubusercontent.com/1530031/67440929-da862700-f5c8-11e9-857a-8ae07f16ce7f.png)

If I use `https`, I get an error atm. I'm assuming it's because it's dev which is why I never mentioned it and I'm thinking production and staging will be different than dev.

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+  //A description of the error
+  public let description: String?
+  
+  init(code: Int32, description: String? = nil) {
```
All `CryptographyError` return code = -1. I think we could hardcode it, or at least add as a default init parameter

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+    var error: Unmanaged<CFError>?
+    let signature = SecKeyCreateSignature(key,
+                                          .ecdsaSignatureMessageX962SHA256,
+                                          message.data(using: .utf8)! as CFData,
```
Unwrap and throw error if can't be unwrapped

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+  
+  /// Determines if a key exists in the keychain without retrieving it
+  public class func keyExists(id: String) -> Bool {
+    let query: [CFString: Any] = [
```
This query is the same as in `isKeyRequiringBiometrics` perhaps we could extract it?

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+  public class func keyExists(id: String) -> Bool {
+    let query: [CFString: Any] = [
+      kSecClass: kSecClassKey,
+      kSecAttrApplicationTag: id.data(using: .utf8)!,
```
Remove force cast

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+    var result: CFTypeRef?
+    let error = SecItemCopyMatching(query as CFDictionary, &result)
+    return error == errSecSuccess || error == errSecInteractionNotAllowed
+      && result != nil && CFDictionaryGetCount((result! as! CFDictionary)) > 0 //swiftlint:disable:this force_cast
```
Remove force cast if it's better to fail safely

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+  public class func getExistingKey(id: String) throws -> CryptographicKey? {
+    let query: [CFString: Any] = [
+      kSecClass: kSecClassKey,
+      kSecAttrApplicationTag: id.data(using: .utf8)!,
```
Remove force unwrapping

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,240 @@
+    let error = SecItemCopyMatching(query as CFDictionary, &result)
+    if error == errSecSuccess || error == errSecDuplicateItem || error == errSecInteractionNotAllowed {
+      if let res = result {
+        return CryptographicKey(key: res as! SecKey, keyId: id) //swiftlint:disable:this force_cast
```
remove force cast

iccub reviewed (changes requested):
Mostly some concerns about force casting and force unwrapping, I don't think we should use so many force casts in this code.
If there is a reason for it please let me know

Brandon-T committed:
Send "ASN.1 DER - SHA256 Fingerprint" to the server in the query parameters for attestation, instead of ASN.1 DER PEM formatted key.

Brandon-T committed:
Removed deprecated SHA1 algorithm as we decided 100% on SHA256.

Brandon-T committed:
Abstracted away boiler plate code into a function.

iccub commented on Client/DeviceCheck/DeviceCheck.swift:
```diff
@@ -0,0 +1,371 @@
+  public func getAttestation(paymentId: String, _ completion: @escaping (AttestationBlob?, Error?) -> Void) {
+    do {
+      guard let privateKey = try Cryptography.getExistingKey(id: DeviceCheckClient.privateKeyId) else {
+        throw CryptographyError(code: -1, description: "Unable to retrieve existing private key")
```
Remove `code`

iccub commented on Client/DeviceCheck/DeviceCheck.swift:
```diff
@@ -0,0 +1,371 @@
+      }
+      
+      guard let publicKey = try privateKey.getPublicKeySha256FingerPrint() else {
+        throw CryptographyError(code: -1, description: "Unable to retrieve public key")
```
Remove `code`

iccub commented on Client/DeviceCheck/DeviceCheck.swift:
```diff
@@ -0,0 +1,371 @@
+  public func setAttestation(nonce: String, _ completion: @escaping (Error?) -> Void) {
+    do {
+      guard let privateKey = try Cryptography.getExistingKey(id: DeviceCheckClient.privateKeyId) else {
+        throw CryptographyError(code: -1, description: "Unable to retrieve existing private key")
```
Remove `code`

iccub commented on Client/DeviceCheck/DeviceCheck.swift:
```diff
@@ -0,0 +1,371 @@
+  public func generateEnrollment(paymentId: String, token: String, _ completion: (DeviceCheckRegistration?, Error?) -> Void) {
+    do {
+      guard let privateKey = try Cryptography.generateKey(id: DeviceCheckClient.privateKeyId) else {
+        throw CryptographyError(code: -1, description: "Unable to generate private key")
```
Remove `code`

iccub commented on Client/DeviceCheck/DeviceCheck.swift:
```diff
@@ -0,0 +1,371 @@
+      }
+      
+      guard let publicKey = try privateKey.getPublicAsPEM() else {
+        throw CryptographyError(code: -1, description: "Unable to retrieve public key")
```
Remove `code`

iccub commented on Client/DeviceCheck/DeviceCheck.swift:
```diff
@@ -0,0 +1,371 @@
+    urlComponents.host = url.host
+    urlComponents.path = url.path
+    urlComponents.percentEncodedQuery = query
+    //urlComponents.queryItems = parameters.map({
```
Is this code still needed?

iccub commented on Client/DeviceCheck/DeviceCheck.swift:
```diff
@@ -0,0 +1,371 @@
+}
+
+public struct DeviceCheckEnrollment: Codable {
+  // Not sure yet
```
Still not sure or can this comment be updated?

iccub commented on Client/DeviceCheck/Cryptography.swift:
```diff
@@ -0,0 +1,279 @@
+        let item = result[kSecAttrAccessControl as String] as CFTypeRef?,
+        CFGetTypeID(item) == SecAccessControlGetTypeID() {
+        
+        //cbio, pbio (.currentBioSet, .presentBioSet)
```
remove if not needed

iccub reviewed (changes requested):
Looks good in general, need some cleanup - removing commented out code

Brandon-T committed:
Cleanup and clarification of comments.

Brandon-T committed:
Removed unused code.

iccub reviewed (approved):
Goood, do we wait with this PR until server side is complete?

Brandon-T committed:
Removed manual encoding as the server no longer uses base64 in the query parameter. We now use hex instead so it's ok to rely on the Apple implementation.

Brandon-T commented:
> Goood, do we wait with this PR until server side is complete?

The server side works with this PR right now. We're currently waiting on grants to be completed but I'm working on integrating this into the Wallet creation atm (in a local branch) & waiting on Security team to give the ok for this PR. Once it's merged, we can completely integrate it into wallet and make a PR for that. Then we wait again for the grant stuff to be complete & continue integration.

jseidl reviewed (approved):
Reviewed the code for common general programming issues but @riastradh-brave is a better suited person for attesting the cryptography-related code.

All looks good for me. Good job folks!

Brandon-T committed:
Added PublicKeyFingerprint as HEX encoded string.
Fixed bug with encoding of the JSON signature. For some stupid reason, if I do NOT specify sortedKeys, it calculates a different signature every single time even if the fields are the SAME. Also some issue with backslash encoding has been fixed.
I have removed backslash encoding in the previous PR.
Added verification of a signature.

Brandon-T committed:
Added HTTPS base URL.

Brandon-T committed:
Removed test code :l

jumde reviewed (approved):
lgtm
